### PR TITLE
DAOS-11076 control: Ensure PMem block device name matches socket ID

### PR DESCRIPTION
**Pull request by tanabarr**
After resetting SCM with daos_server storage prepare command, the PMem
block device names did not match up with the socket ID.

Ensure that when creating PMem namespaces with daos_server tool the ID
defined in the block device name matches the socket ID of the underlying
region.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>
Required-githooks: true

### CODE REVIEW

File: src/control/server/storage/scm/ipmctl_region.go [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+	regionsPerSocket := make(socketRegionMap)
+	for _, region := range rl.Regions {
+		sockID := int(region.SocketID)
+		if _, exists := regionsPerSocket[sockID]; exists {
+			return errors.Errorf("unexpected second region assigned to socket %d", sockID)
+		}
+		regionsPerSocket[sockID] = region
+	}
+	*srm = regionsPerSocket
```
Comment (mjmac) [inline src/control/server/storage/scm/ipmctl_region.go:318]: I don't think this is necessary, as maps in Go are [already reference types](https://go.dev/blog/maps). You should be able to use a non-pointer receiver for the method and just modify the map directly.

https://go.dev/play/p/FNtZyLnq4La

File: src/control/server/storage/scm/ipmctl_region.go [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+	}
+
+	keys := make([]int, 0, len(*srm))
+	for k := range *srm {
+		keys = append(keys, k)
+	}
+	sort.Ints(keys)
+
+	return keys
+}
```
Comment (mjmac) [inline src/control/server/storage/scm/ipmctl_region.go:293]: This will be changed to a non-pointer receiver, right?

File: src/control/server/storage/scm/ipmctl_region.go [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
+func (srm *socketRegionMap) isEmpty() bool {
```
Comment (mjmac) [inline src/control/server/storage/scm/ipmctl_region.go:295]: This method isn't necessary. Just use `len()` on the map.

**Review by mjmac — COMMENTED** [review]
Took an initial pass through. Seems generally like a good restructuring to add support for per-socket operations.

File: src/control/lib/ipmctl/nvm_types.go [hunk 4]
```diff
@@ -99,6 +101,15 @@ func (pmrt PMemRegionType) String() string {
 	}[pmrt]
 }
 
+func PMemRegionTypeFromString(in string) PMemRegionType {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/ipmctl/nvm_types.go:104]: done

File: src/control/lib/ipmctl/nvm_types.go [hunk 5]
```diff
@@ -115,14 +126,24 @@ const (
...
-		RegionHealthNormal:  "Normal",
+		RegionHealthNormal:  "Healthy",
 		RegionHealthError:   "Error",
 		RegionHealthUnknown: "Unknown",
 		RegionHealthPending: "Pending",
 		RegionHealthLocked:  "Locked",
 	}[pmrh]
 }
 
+func PMemRegionHealthFromString(in string) PMemRegionHealth {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/ipmctl/nvm_types.go:137]: done

File: src/control/server/storage/faults.go [hunk 6]
```diff
@@ -14,30 +14,47 @@ import (
...
 // mode, this is unsupported.
-var FaultScmNotInterleaved = storageFault(code.ScmNotInterleaved,
-	"PMem regions are in AppDirect non-interleaved mode",
-	"Rerun the command first with the --reset option and then without to recreate the "+
-		"PMem regions in the recommended mode")
+func FaultScmNotInterleaved(sockID uint) *fault.Fault {
+	return storageFault(
+		code.ScmBadRegion,
+		fmt.Sprintf("PMem region on socket %d is in non-interleaved mode which is unsupported", sockID),
+		"Remove and recreate region in AppDirect interleaved mode")
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/faults.go:23]: done, @kjacque we only remove the old goals so that when the command is invoked any stale goals won't interfere with any new goals that are set as part of the state change. The process of removing regions requires new goals to be set (after any stale ones have been cleared) before a reboot.

File: src/control/server/storage/faults.go [hunk 7]
```diff
@@ -14,30 +14,47 @@ import (
...
+}
+
+// FaultScmNotHealthy create a fault for the case where the PMem region is in an unhealthy state.
+func FaultScmNotHealthy(sockID uint) *fault.Fault {
+	return storageFault(
+		code.ScmBadRegion,
+		fmt.Sprintf("PMem region on socket %d is unhealthy", sockID),
+		"Check persistent memory modules are not faulty and then remove and recreate region")
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/faults.go:31]: done

File: src/control/server/storage/faults.go [hunk 8]
```diff
@@ -14,30 +14,47 @@ import (
...
+}
+
+// FaultScmPartialCapacity creates a fault for the case where the PMem region has only partial
+// free capacity, this is unsupported.
+func FaultScmPartialCapacity(sockID uint) *fault.Fault {
+	return storageFault(
+		code.ScmBadRegion,
+		fmt.Sprintf("PMem region on socket %d only has partial capacity free", sockID),
+		"Creating namespaces on regions with partial free-capacity is unsupported, remove "+
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/faults.go:40]: done

File: src/control/server/storage/scm/ipmctl_region.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/storage/scm/ipmctl_region.go:318]: done

File: src/control/server/storage/scm/ipmctl_region.go (hunk 2 above)
Comment (tanabarr) [reply src/control/server/storage/scm/ipmctl_region.go:293]: done

File: src/control/server/storage/scm/ipmctl_region.go (hunk 3 above)
Comment (tanabarr) [reply src/control/server/storage/scm/ipmctl_region.go:295]: done

File: src/control/lib/ipmctl/nvm_types.go [hunk 9]
```diff
@@ -115,14 +126,24 @@ const (
 
 func (pmrh PMemRegionHealth) String() string {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/ipmctl/nvm_types.go:127]: done

File: src/control/server/storage/scm.go [hunk 10]
```diff
@@ -26,33 +26,45 @@ type ScmState int
...
-	// ScmStateNoFreeCapacity indicates that PMem modules exist with configured regions and no free
-	// capacity.
-	ScmStateNoFreeCapacity
-	// ScmStateNotInterleaved indicates that PMem modules exist with configured regions but not
-	// in AppDirect interleaved/persistent mirror mode.
-	ScmStateNotInterleaved
-	// ScmStateNoModules indicates that no PMem modules exist.
-	ScmStateNoModules
+	// NoRegions indicates that PMem modules exist, but no regions have been created.
+	ScmNoRegions
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm.go:30]: done

File: src/control/server/storage/scm/provider.go [hunk 11]
```diff
@@ -276,58 +276,76 @@ func NewProvider(log logging.Logger, backend Backend, sys SystemProvider) *Provi
...
+	if req.SocketID != nil {
+		sockSelector = int(*req.SocketID)
+	}
+
+	modules, err := p.backend.getModules(sockSelector)
 	if err != nil {
 		return nil, err
 	}
-	p.log.Debugf("scm backend: %d modules", len(modules))
+	msg = fmt.Sprintf(": %d pmem modules", len(modules))
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm/provider.go:306]: yes, mistake fixed in subsequent commit

File: src/control/server/storage/scm/provider.go [hunk 12]
```diff
@@ -276,58 +276,76 @@ func NewProvider(log logging.Logger, backend Backend, sys SystemProvider) *Provi
...
 
 	// If there are no modules, don't bother with the rest of the scan.
 	if len(modules) == 0 {
-		return &storage.ScmScanResponse{
-			State: storage.ScmStateNoModules,
-		}, nil
+		resp.State.State = storage.ScmNoModules
+		msg = fmt.Sprintf(": pmem state %q", resp.State.State)
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm/provider.go:311]: yes, mistake fixed in subsequent commit

File: src/control/server/storage/scm/provider.go [hunk 13]
```diff
@@ -276,58 +276,76 @@ func NewProvider(log logging.Logger, backend Backend, sys SystemProvider) *Provi
...
 		return nil, err
 	}
-	p.log.Debugf("scm backend: namespaces %+v", namespaces)
+	resp.State = *sockState
+	msg = fmt.Sprintf(": pmem state %+v", *sockState)
 
-	state, err := p.backend.getRegionState()
-	if err != nil {
-		return nil, err
+	if sockState.State != storage.ScmNoRegions {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm/provider.go:323]: re-factored in subsequent commit to make this code obsolete

File: src/control/server/storage/scm/ipmctl.go [hunk 14]
```diff
@@ -248,219 +99,173 @@ func (cr *cmdRunner) getModules() (storage.ScmModules, error) {
...
+		case storage.ScmFreeCap:
+			fc := humanize.Bytes(uint64(r.FreeCapacity))
+			cr.log.Debugf("socket %d app-direct region has %s free", r.SocketID, fc)
+			hasFreeCap = true
+
+		case storage.ScmNoFreeCap:
+			// Fall-through
+
+		default:
+			return nil, errors.New("unexpected state from getRegionState")
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm/ipmctl.go:148]: done

**Review by kjacque — APPROVED** [review]
My comments are all minor things that could be addressed in a follow-on.

File: src/control/server/storage/faults.go [hunk 15]
```diff
@@ -14,30 +14,48 @@ import (
...
+		fmt.Sprintf("Refer to the ipmctl instructions in the troubleshooting section of the DAOS admin guide to check PMem module health and replace faulty PMem modules. %s", recreateRegionsStr))
+}
+
+// FaultScmPartialCapacity creates a fault for the case where the PMem region has only partial
+// free capacity, this is unsupported.
+func FaultScmPartialCapacity(sockID uint) *fault.Fault {
+	return storageFault(
+		code.ScmBadRegion,
+		fmt.Sprintf("PMem region on socket %d only has partial capacity free", sockID),
+		"Creating namespaces on regions with partial free-capacity is unsupported, remove namespaces by running the command with --reset and then without --reset, no reboot should be required")
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/faults.go:42]: The command surface change will enable a user to act on a single socket at a time, there is only ever one region per socket and the design is such that the command will use all the capacity of a region, running the command when a region has only partial capacity will still not be supported.

File: src/control/cmd/daos_server/storage.go [hunk 16]
```diff
@@ -139,20 +139,20 @@ func (cmd *storagePrepareCmd) prepScm(scanErrors *errs, prep scmPrepFn) error {
 	}
 	cmd.Debugf("%s scm resp: %+v", op, resp)
 
-	state := resp.State
+	state := resp.State.State
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/cmd/daos_server/storage.go:142]: done

File: src/control/server/storage/scm/ipmctl_namespace.go [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+// For each region, create <nrNsPerSocket> namespaces.
+func (cr *cmdRunner) createNamespaces(regionsPerSocket socketRegionMap, nrNsPerSock uint) error {
+	if nrNsPerSock < minNrNssPerSocket || nrNsPerSock > maxNrNssPerSocket {
+		return errors.Errorf("unexpected number of namespaces requested per socket: want [%d-%d], got %d",
+			minNrNssPerSocket, maxNrNssPerSocket, nrNsPerSock)
+	}
+
+	nrRegions := len(regionsPerSocket)
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm/ipmctl_namespace.go:58]: there is a 1:1 mapping between socket and region, so technically it's either, I will make the change anyway

File: src/control/server/storage/scm/ipmctl_namespace.go [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+	if nrRegions == 0 {
+		return errors.New("unexpected number of pmem regions (0)")
+	}
+	cr.log.Debugf("creating %d namespaces per socket (%d regions)", nrNsPerSock, len(regionsPerSocket))
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm/ipmctl_namespace.go:62]: as above

File: src/control/server/storage/scm/ipmctl.go [hunk 19]
```diff
@@ -248,219 +99,194 @@ func (cr *cmdRunner) getModules() (storage.ScmModules, error) {
...
+
+		min, err := strconv.ParseUint(matches[2], 16, 10)
+		if err != nil {
+			return errors.Wrapf(err, "parse minor namespace version (%q)", ns.Name)
+		}
+
+		log.Debugf("found namespace %d.%d on numa %d", maj, min, ns.NumaNode)
+
+		nsMajMinMap[int(maj)] = append(nsMajMinMap[int(maj)], int(min))
+		sort.Ints(nsMajMinMap[int(maj)])
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm/ipmctl.go]: done

File: src/control/server/storage/scm/ipmctl.go [hunk 20]
```diff
@@ -248,219 +99,173 @@ func (cr *cmdRunner) getModules() (storage.ScmModules, error) {
...
-
-		kv := strings.Split(entry, "=")
-		if len(kv) != 2 {
-			continue
+	for sid, region := range regionsPerSocket {
+		// Check value is 2MiB aligned and (TODO) multiples of interleave width.
+		pmemBytes := uint64(region.FreeCapacity) / uint64(nrNsPerSocket)
+
+		if pmemBytes%alignmentBoundaryBytes != 0 {
+			return nil, errors.Errorf("socket %d: free region size is not 2MiB aligned", sid)
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm/ipmctl.go]: done

**Review by mjmac — APPROVED** [review]
Some quibbles about the fault strings, but I won't block landing on them. How has this PR been tested? Have you verified that a system's pmem can be fully reset/recreated (i.e. gone through the full process with multiple reboots, etc)?

File: src/control/server/storage/faults.go [hunk 21]
```diff
@@ -14,30 +14,56 @@ import (
...
+}
+
+// FaultScmUnknownMemoryMode creates a Fault for the case where the PMem region has an unsupported
+// persistent memory type (not AppDirect).
+func FaultScmUnknownMemoryMode(sockID uint) *fault.Fault {
+	return storageFault(
+		code.ScmBadRegion,
+		fmt.Sprintf("PMem region on socket %d has an unsupported persistent memory type", sockID),
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/faults.go:50]: - [ ] fix in follow-up PR https://github.com/daos-stack/daos/pull/9541

**Comment by tanabarr** [comment]
> Some quibbles about the fault strings, but I won't block landing on them. How has this PR been tested? Have you verified that a system's pmem can be fully reset/recreated (i.e. gone through the full process with multiple reboots, etc)?

Yes I have tested the PR on wolf-157 building from source, reset and recreating PMem with reboots. Have also been doing the same on wolf-220 with RPM installation. No functional regressions found.

**Comment by tanabarr** [comment]
> tl;dr: Let's keep these strings short & sweet with high information density.

- [ ] revise fault text with the guidance in this PRs comments in follow-up PR https://github.com/daos-stack/daos/pull/9541

> As an aside, I wonder if we should standardize on either "PMem" or "SCM" across all of our command surface and documentation. I think it's potentially confusing to new users to have multiple terms for the same thing. That's something we can discuss in a UX WG.

- [ ] Change command syntax from "scm" to "pmem" in follow-up PR https://github.com/daos-stack/daos/pull/9541

**Review by knard-intel — APPROVED** [review]
Seems OK for me for what I understand

File: src/control/server/storage/scm/ipmctl.go [hunk 22]
```diff
@@ -7,39 +7,31 @@
...
+	"github.com/daos-stack/daos/src/control/fault"
 	"github.com/daos-stack/daos/src/control/lib/ipmctl"
 	"github.com/daos-stack/daos/src/control/logging"
 	"github.com/daos-stack/daos/src/control/server/storage"
 )
 
 const (
 	minNrNssPerSocket = 1
-	maxNrNssPerSocket = 8
+	maxNrNssPerSocket = 4
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/storage/scm/ipmctl.go:27]: Mainly because we don't anticipate an immediate need/usecase but also because there hasn't been any testing of higher values and don't want to give the impression that we have validated using higher values.
